File: solutions/1466/solution.ml
let createGraph n connections =
  let graph = Array.make n [] in

  List.iter
    (fun (a, b) ->
      graph.(a) <- (b, 1) :: graph.(a);
      graph.(b) <- (a, 0) :: graph.(b))
    connections;

  graph

let minReorder n connections =
  let graph = createGraph n connections in

  let visited = ref [] in

  let res = ref 0 in

  let rec dfs city graph visited =
    visited := city :: !visited;

    List.iter
      (fun (neighbor, reverse) ->
        if not (List.mem neighbor !visited) then (
          res := !res + reverse;
          dfs neighbor graph visited))
      graph.(city)
  in

  dfs 0 graph visited;
  !res

File: solutions/1466/solution.py
class Solution:
    def minReorder(self, n, connections):
        graph = self.createGraph(n, connections)
        visited = set()
        self.res = 0
        self.dfs(0, graph, visited)

        return self.res

    def createGraph(self, n, connections):
        graph = [[] for _ in range(n)]

        for a, b in connections:
            graph[a].append((b, 1))
            graph[b].append((a, 0))

        return graph

    def dfs(self, city, graph, visited):
        visited.add(city)

        for neighbor, reverse in graph[city]:
            if neighbor not in visited:
                self.res += reverse
                self.dfs(neighbor, graph, visited)
